### ISSUE TITLE
Document using site-wide JSX pragma configuration

### ISSUE THREAD

**Issue by lachlanjc** [system-ui/theme-ui#1335]
```json
{
  "plugins": [
    [
      "@babel/plugin-transform-react-jsx",
      {
        "throwIfNamespace": false, // defaults to true
        "runtime": "automatic", // defaults to classic
        "importSource": "theme-ui" // defaults to react
      }
    ]
  ]
}
```

https://babeljs.io/docs/en/babel-plugin-transform-react-jsx

**Comment by adamsoffer** [comment]
@lachlanjc I'm still getting this error `Property 'sx' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'` when using the sx prop on a regular div element. I'm on the latest version (0.6.0-alpha.4). Any idea why?

My .babelrc file looks like this: 
```json
{
  "presets": ["next/babel"],
  "plugins": [
    "inline-react-svg",
    [
      "@emotion/babel-plugin-jsx-pragmatic",
      {
        "module": "theme-ui",
        "import": "jsx",
        "export": "jsx"
      }
    ],
    [
      "@babel/plugin-transform-react-jsx",
      {
        "throwIfNamespace": false, // defaults to true
        "runtime": "automatic", // defaults to classic
        "importSource": "theme-ui" // defaults to react
      }
    ]
  ]
}
```

**Comment by lachlanjc** [comment]
Eek! I haven't tried using this myself so I have no idea how it works in practice, just opened this issue for the future. I only use the components on my sites.

**Comment by hasparus** [comment]
okay @adamsoffer, I mostly use `sx` and always TypeScript :D Feel free to `@` me when you encounter problems like this.

Thanks for this comment! That's actually something we need to cover in the docs.

---

Previously, `sx` was added to _global JSX types_, so it was super easy to conflict with other libraries.
Right now, we're exporting JSX namespace, and TypeScript adds `sx` prop when it sees the pragma comment.

That's nice, because we're not polluting global JSX namespace anymore, but this means that if you configure Babel to support `sx` everywhere you need to add `sx` manually to React types.

You need to add the following to one of your files.

```ts
// To add `css` if you're using it somewhere
/// <reference types="@emotion/react/types/css-prop" />

import {} from 'react'
import { ThemeUIStyleObject } from '@theme-ui/core' // or 'theme-ui'

declare module 'react' {
  interface Attributes {
    sx?: ThemeUIStyleObject
  }
}
```

- Emotion 11 Migration Notes describe how to add `css` prop globally — https://emotion.sh/docs/emotion-11#css-prop-types
- TypeScript 3.0 Release Notes explain what's happening underneath — https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#support-for-defaultprops-in-jsx

---

We'll add an entry point similar to `@emotion/react/types/css-prop` before a stable release, so it's going to be a bit more convenient to use.

**Comment by adamsoffer** [comment]
ah thanks for this @hasparus !

**Comment by flo-sch** [comment]
I am struggling a bit with those pragma changes, seeking for help here :D

I am currently using theme-ui@0.4.0-rc.14 (React v16.13.1, Emotion v10, classic pragma).

I use theme-ui:
- directly inside projects, compiled with webpack & babel
- through a UI library with additional custom components
  - currently full of `/** @jsx jsx */` statements everywhere (which I would like to get rid of if possible)
  - compiled from TS with microbundle v0.13 (`--jsx React.createElement`) --> no babel config!
  - peerDependency to react (>=16.8 <16.14) & theme-ui (^0.4.0-rc.14)
  - documented with Storybook (which I think released opt-in Emotion 11 support in its v6.2 from what I understand)
(the projects I am talking about use both theme-ui and this UI library)

Yesterday I had to hotfix an error in production caused by a sneaky misplaced import which broke my design-system package in one project:

```diff
// some react component inside my UI library project
/** @jsx jsx */
// Adding the following import by mistake broke the interwebs
+ import React from 'react';
import { Flex, jsx } from 'theme-ui';
``` 

> Uncaught ReferenceError: React is not defined

For this reason among others, I plan to sync a migration of React v17.0, Emotion v11.0 and theme-ui v0.6.0
In my projects, I would like to configure the pragma site-wise for both theme-ui and my UI library.
I plan to achieve that through a custom babel config with automatic jsx pragma from `@babel/preset-react`:
https://babeljs.io/docs/en/babel-preset-react

What remains unclear to me is mainly my UI library part, despite obviously bumping its peerDependencies:
- should I build it in a specific way for classic & automatic pragma to work?
- Is that possible with microbundle v0.13? (I've seen some ongoing work to move theme-ui to preconstruct instead, wondering if this is related?)
  - found [some dark magic here](https://github.com/developit/microbundle/issues/763#issuecomment-778848944), not sure if that would help?
- do I still need to `/** @jsx jsx */` in that library, and/or `import React from 'react';` instead?

**Comment by hasparus** [comment]
Hey @flo-sch. Nice to see you again!

> Is that possible with microbundle v0.13? (I've seen some ongoing work to move theme-ui to preconstruct instead, wondering if this is related?)

Oh that's finished already, and we just had a problem with microbundle build times. It shouldn't affect users at all.

> What remains unclear to me is mainly my UI library part, despite obviously bumping its peerDependencies:

If you want to get rid of `/** @jsx jsx */` you can configure `theme-ui` as [`importSource`](https://babeljs.io/docs/en/babel-preset-react#importsource) for your UI library. You could do it in a Babel config or possibly (not 100% sure) with a combination of CLI flags from the comment you linked.

I'd try something like this, but no guarantees:

```
microbundle --jsx jsx --jsxImportSource theme-ui --globals react/jsx-runtime=jsx
```

**Comment by flo-sch** [comment]
Hey @hasparus, thanks, same for me :)

And thanks for you answers!

> If you want to get rid of /** @jsx jsx */ you can configure theme-ui as importSource for your UI library. You could do it in a Babel config or possibly (not 100% sure) with a combination of CLI flags from the comment you linked.

That does work, at least it changes what is being compiled. I am still not sure is how it should be though 😅 

```tsx
// src/components/MyComponent/MyComponent.tsx
import React from 'react';
import { Flex, FlexProps, Heading, HeadingProps } from 'theme-ui';

export interface MyComponentProps extends FlexProps {
  headingProps?: HeadingProps;
}

const MyComponent = React.forwardRef<HTMLDivElement, FlexProps>((({ children, headingProps, ...props }), ref) => (
  <Flex {...props} ref={ref}>
    <Heading {...headingProps}>Title</Heading>
    {children}
  </Flex>
));
```

If I compile it with `microbundle build --jsx React.createElement`, then I get (esm version):

```js
import React from 'react';
import { Flex, Heading, Text } from 'theme-ui';

// ...

var MyComponent = React.forwardRef(function (_ref, ref) {
  var children = _ref.children,
      headingProps = _ref.headingProps,
      props = _objectWithoutPropertiesLoose(_ref, ["children", "headingProps"]);

  return /*#__PURE__*/React.createElement(Flex, _extends({}, props, {
    ref: ref
  }), /*#__PURE__*/React.createElement(Heading, headingProps, "Title"), children);
});
```

If I compile it with `microbundle build --jsx jsx --jsxImportSource theme-ui --globals react/jsx-runtime=jsx` then I get:

```diff
import React from 'react';
import { Flex, Heading, Text } from 'theme-ui';
+ import { jsxs, jsx } from 'theme-ui/jsx-runtime';

// ...

- var MyComponent = React.forwardRef(function (_ref, ref) {
+ var MyComponent = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var children = _ref.children,
      headingProps = _ref.headingProps,
      props = _objectWithoutPropertiesLoose(_ref, ["children", "headingProps"]);

-  return /*#__PURE__*/React.createElement(Flex, _extends({}, props, {
-    ref: ref
-  }), /*#__PURE__*/React.createElement(Heading, headingProps, "Title"), children);
+  return jsxs(Flex, _extends({}, props, {
+    ref: ref,
+    children: [jsx(Heading, _extends({}, headingProps, {
+      children: "Title"
+    })), children]
+  }));
});
```

Locally, with theme-ui and react installed as devDependencies, both of those compilations seem to work.

I am less sure how they will behave once installed in my "end-project" though, in which I should have a babel config to configure the jsx runtime (which should not apply to node_modules I guess?)

If I want my UI library to allow for both classic & automatic pragma in my end-project, which runtime should I compile it with?
(Or perhaps it doesn't even matter?)

**Comment by hasparus** [comment]
> If I want my UI library to allow for both classic & automatic pragma in my end-project, which runtime should I compile it with?

Well, both pragmas in your library will get compiled away, so they shouldn't affect the end-project in any way. All functions behind the pragma are "compatible" with each other if you're worried about it.

**Comment by flo-sch** [comment]
Alright, that makes sense to read!
For some reason I guess the bug I faced is what got me worried about that in the first place 😅 

Where do you think this should be documented? Perhaps I can gather something?

**Comment by hasparus** [comment]
We might need a separate page in the docs describing what JSX pragma is, linking to original sources (like React and Babel docs) and listing ways of configuring Theme UI with it.

@lachlanjc?

**Comment by lachlanjc** [comment]
We have this: https://theme-ui.com/guides/jsx-pragma

What could we add/how could we make it better?

**Comment by flo-sch** [comment]
I guess how to configure it with (and without) Babel, TS and/or bundlers would be nice 🤔

I am not thinking about an in-depth guides for each and every bundler of course, since they all work differently, but perhaps we could cover what is needed for both classic and automatic runtime to work?

Also what about `@babel/preset-typescript`, do we need to configure both `tsconfig.json` and/or babel in that case?

**Comment by hasparus** [comment]
Okay, I think I've found something we missed. We should definitely link to `/jsx-pragma` from Getting Started when `sx` prop is shown for the first time — https://theme-ui.com/getting-started#sx-prop.

What do you guys think about something short like:

> _Confused about this `@jsxImportSource` comment? Read more about [JSX Pragma](https://theme-ui.com/guides/jsx-pragma)._

?

**Comment by lachlanjc** [comment]
Great suggestion @hasparus! Implemented: https://github.com/system-ui/theme-ui/commit/b61a7d9f47eb04d2368c6f9ef0c4b7052ab6c06b

**Comment by hasparus** [comment]
Okay, so let's wait a week and see if anybody asks "what's that weird comment?" again and if not then we can close this?

**Comment by lachlanjc** [comment]
No, this issue is larger—we should have docs on setting up the pragma project-wide using Babel config. I'm just not likely to actually add those docs because I don't use the pragma/haven't set this up before

**Comment by flo-sch** [comment]
I could try something, will PR on the pragma page

**Comment by hasparus** [comment]
Fair point, Lachlan. Something like: 

> - If you want, you can set up your transpiler to use Theme UI's JSX pragma everywhere.
> - few examples:
>   - Here's how you configure [Babel Preset React](https://babeljs.io/docs/en/babel-preset-react) [code snippet]
>   - Here's how you configure @babel/plugin-transform-react-jsx [code snippet like in original post]
>   - Here's how you configure in in tsconfig.json if you use TypeScript to transpile your JSX
>   - Here's how you do it in Next.js (_babel.config.js_ with `presets: [["next/babel", { "preset-react": sameAsForPresetReact } ]]`)  
> - Remember, you only need to configure it once. 

?

I noticed I still have some projects with classic runtime configured for global JSX from Theme UI with _babel-plugin-jsx-imports_, but I think we'll be fine with documenting automatic runtime only.

**Comment by lachlanjc** [comment]
Sounds great! Anyone is free to take this task

**Comment by flo-sch** [comment]
I claim it, just helping my parents move out this WE so a bit busy here, but I promise I will early next week the latest :)

**Comment by hasparus** [comment]
No rush @flo-sch :)

**Comment by flo-sch** [comment]
@hasparus the docs today states:

> The preferred way of using these functions is by adding the custom pragma comment to the top of your file.

Is it for performance reasons that we recommend that? (Do we know a bit more about it...?)

**Comment by hasparus** [comment]
> The preferred way of using these functions is by adding the custom pragma comment to the top of your file.

Well, it's certainly preferred to importing them and calling them explicitly, but I didn't write that line if you're asking about it. I've been running as Theme UI's `jsx` and Emotion's `jsx` as a default pragma for a while — you should be fine but I wouldn't say any syntax is "blessed".

**Comment by flo-sch** [comment]
I am mainly just curious to know how it affects performances. I am usually using `sx` everywhere in my projects so automatic runtime is easier for me as well

**Comment by hasparus** [comment]
> I am mainly just curious to know how it affects performances.

I made a benchmark once, but I lost it somewhere. I'll need to get to it some day and add one here with a nice CI integration.

Essentially, every JSX element is transpiled to `jsx` function from Theme UI which shallow copies all props, transforms `sx` if it's there and calls Emotion's `jsx` function... which actually checks if props contain `'css'` prop first... Okay... This looks a tiny bit better than ours :D It might be good to add a benchmark regardless...

**Comment by hasparus** [comment]
:rocket: Issue was released in [`v0.10.0`](https://github.com/system-ui/theme-ui/releases/tag/v0.10.0) :rocket:
